packages/ava/src/ntv/generate/utils.ts

```diff
@@ -0,0 +1,78 @@
+import { get, isArray, isUndefined, slice, isNumber, toString, replace, isString, startsWith } from 'lodash';
+
+import type { Variable, VariableMeta } from './types';
+import type { EntityType } from '../schema';
+
+export type TemplateStructure = {
+  type: 'text' | 'variable' | 'template';
+  value: string;
+};
+
+/**
+ * simple path system
+ * if starts with '.', it's relative path
+ * // TODO 支持相对路径找到上级
+ * if starts with '..', it's can get prev level variable
+ * otherwise, it's absolute path
+ */
+export function getByPath(globalVar: Variable, scopeVar: Variable, path?: string) {
+  if (!path) return scopeVar;
+  if (startsWith(path, '.')) return get(scopeVar, path.slice(1));
+  return get(globalVar, path);
+}
+
+/** parse template string */
+export function templateStr2Structure(templateStr: string): TemplateStructure[] {
+  // eslint-disable-next-line no-useless-escape
+  const splitReg = /([\$|&]{.*?})/;
+  const varReg = /\${(.*?)}/;
+  const tempReg = /&{(.*?)}/;
+  return templateStr
+    .split(splitReg)
+    .filter((str) => str)
+    .map((str) => {
+      const templateId = tempReg.exec(str)?.[1];
+      if (templateId) return { value: templateId, type: 'template' };
+      const varName = varReg.exec(str)?.[1];
+      if (varName) return { value: varName, type: 'variable' };
+      return { value: str, type: 'text' };
+    });
+}
+
+/** format any to array, and filter nil */
+export function formattedAsArray(variable: Variable) {
+  if (isUndefined(variable)) return [];
+  return isArray(variable) ? variable : [variable];
+}
+
+export function getScopeVariableArray(globalVar: Variable, scopeVar: Variable, path: string, limit?: number) {
+  return slice(formattedAsArray(getByPath(globalVar, scopeVar, path)), 0, limit);
+}
+
+export function getAssessment(entityType: EntityType, value: any) {
+  if (!isNumber(value)) return undefined;
+  if (entityType === 'delta_value' || entityType === 'ratio_value') {
+    if (value > 0) return 'positive';
+    if (value < 0) return 'negative';
+  }
+  return undefined;
+}
+
+export function getFormattedNumberValue(entityType: string, value: any, formatter?: VariableMeta['formatter']) {
+  if (!isNumber(value)) return value;
+  if (entityType === 'delta_value' || entityType === 'ratio_value') {
```

Comment:
entityType is string, could it be enumerable? There is a type `EntityType` in L52.
`'delta_value'` and `'ratio_value` don't seem to be prompted by the editor.

Reply:
Changed argument naming to varType, could be entityType or customType